### DISCUSSION TITLE
Implement water_heater.turn_on and water_heater.turn_off methods

### DISCUSSION THREAD

**Discussion by jbouwh** [home-assistant/architecture#928] [Entity Models] [answered]
To be able to implement `water_heater` support in `google_assistant` we need to support for the `water_heater.turn_on` and `water_heater.turn_off` service, but this is not provided by the core `water_heater` component.

See https://developers.home.google.com/cloud-to-cloud/guides/waterheater

The device can return and update the current system operation_mode when it is turned on.

The implementation can very similar to that of `climate` and  might not seem very functional, but is mainly meant to allow enabling this for integrations that are missing this feature right now, as now only the operation_mode can be set, hence the device cannot just be turned on.

PR to implement this suggestion:
https://github.com/home-assistant/core/pull/94817

**Comment by emontnemery** [comment]
The implementation in climate is not great because it just picks an operation mode supported by the water heater.
Instead of copying that pattern, should we add `turn_on` + `turn_off` services to `water_heater` with no default implementation together with feature flags?

**Reply by frenck** [reply]
It is a `ToggleEntity`, unless we add such support to the `ToggleEntity` (and all the child), I don't think a support flag is right here.

(The `ToggleEntity` s*cks in a way though).

I don't think the current (default) implementation of climate is that bad though?

**Reply by jbouwh** [reply]
> It is a `ToggleEntity`, unless we add such support to the `ToggleEntity` (and all the child), I don't think a support flag is right here.
> 
> (The `ToggleEntity` s*cks in a way though).
> 
> I don't think the current (default) implementation of climate is that bad though?

For water_heater the list of default modes to consider when seem limited:
[ STATE_HEAT_PUMP, STATE_GAS, STATE_ELECTRIC ]
Chances are low that more than one of these modes are implemented simultaneously (but it is possible of course). So the default `turn_on` experience should not be that bad.
It is to the integrations after all to report back the correct actual operation mode after a device has been turned on. This strictly is not something to implement in the `water_heater.turn_on` service. It is more an issue that we would like to have the feature of an optimistic state, because a device lacks to report.
So, the conclusion for a plan-B: We also could raise a `NotImplementedError` as the default behavior, and not copy the climate code here.

**Reply by emontnemery** [reply, marked as answer]
As discussed yesterday:
- `WaterHeaterEntity` is not a `ToggleEntity`
- While the default implementation of `climate.turn_on` is not that bad it's also not that great, and there's no reason to repeat it here

Hence, let's add `turn_on` + `turn_off` services to `water_heater` with no default implementation together with feature flags

**Reply by frenck** [reply]
implemented in https://github.com/home-assistant/core/pull/94817
